lib/core/src/Cardano/Wallet/Network/Light.hs

```diff
@@ -0,0 +1,237 @@
+    -- ^ We are entering the unstable region.
+    -- @Unstable blocks new tip@.
+
+boldyGoWhereNextPoint
+    :: Monad m
+    => LightSyncSource m block addr txs
+    -> ChainPoint
+    -> m (NextPointAdventure block)
+boldyGoWhereNextPoint light pt = do
+    tip <- getTip light
+    mold <- getBlockHeaderAt light pt
+    maybeRollback mold $ \old ->
+        if isUnstable (stabilityWindow light) old tip
+        then do
+            mblocks <- getNextBlocks light $ chainPointFromBlockHeader old
+            maybeRollback mblocks $ \case
+                [] -> pure $ Unstable [] old tip
+                (b:bs) -> do
+                    let new = getHeader light $ NE.last (b :| bs)
+                    continue <- isConsensus light $ chainPointFromBlockHeader new
+                    pure $ if continue
+                        then Unstable (b:bs) new tip
+                        else Rollback
+        else do
+            mnew <- getBlockHeaderAtHeight light $
+                blockHeightToInteger (blockHeight tip) - stabilityWindow light
+            maybeRollback mnew $ \new -> pure $ Stable old new tip
+  where
+    maybeRollback m f = maybe (pure Rollback) f m
+
+-- | Test whether a 'ChainPoint' is in the
+-- unstable region close to the tip.
+isUnstable :: BlockHeight -> BlockHeader -> BlockHeader -> Bool
+isUnstable stabilityWindow_ old tip =
+    blockHeightToInteger (blockHeight tip) - stabilityWindow_
+  <= blockHeightToInteger (blockHeight old)
+
+secondsPerSlot :: DiffTime
+secondsPerSlot = 2
+
+-- | Create a 'BlockSummary' 
+mkBlockSummary
+    :: LightSyncSource m block addr txs
+    -> BlockHeader
+    -> BlockHeader
+    -> BlockSummary m addr txs
+mkBlockSummary light old new = BlockSummary
+    { from = old
+    , to = new
+    , query = getAddressTxs light old new
+    }
+
+blockHeightToInteger :: Quantity "block" Word32 -> Integer
+blockHeightToInteger (Quantity n) = fromIntegral n
+
+{-------------------------------------------------------------------------------
+    Logging
+-------------------------------------------------------------------------------}
+data LightLayerLog
+    = MsgLightRollForward ChainPoint BlockHeader BlockHeader
+    | MsgLightRollBackward ChainPoint ChainPoint
+    deriving (Show, Eq, Generic)
+
+instance ToText LightLayerLog where
+    toText = \case
+        MsgLightRollForward from_ to_ tip -> T.unwords
+            [ "LightLayer roll forward:"
+            , "from: ", toText $ show from_
+            , "to: ", toText $ show to_
+            , "tip: ", toText $ show tip
+            ]
+        MsgLightRollBackward from_ to_ -> T.unwords
+            [ "LightLayer roll backward:"
+            , "from: ", toText $ show from_
+            , "to: ", toText $ show to_
+            ]
+
+instance HasPrivacyAnnotation LightLayerLog
+
+instance HasSeverityAnnotation LightLayerLog where
+    getSeverityAnnotation = \case
+        MsgLightRollForward{} -> Debug
```

Comment:
do we have the same severity when rollback happens in "normal" mode?

Reply:
Yes, for the underlying `ChainSync` protocol:

```
instance HasSeverityAnnotation (ChainSyncLog block point) where
    getSeverityAnnotation = \case
        MsgChainFindIntersect{} -> Debug
        MsgChainRollForward{} -> Debug
        MsgChainRollBackward{} -> Debug
        MsgChainTip{} -> Debug
        MsgLocalTip{} -> Debug
        MsgTipDistance{} -> Debug
```

The wallet has additional logging top of that, `MsgDiscoveredTxs` that notifies of new transactions has severity `Info`.